src/encoder/encoder.cpp

```diff
@@ -15,27 +15,121 @@
+#include "recording/defs_recording.h"
+// TODO: Note: this is incomplete, since encoderffmpegmp3.cpp (and vorbis)
+// are not compiled in any case.
+#ifdef __FFMPEGFILE__
+#include "encoder/encoderffmpegmp3.h"
+#include "encoder/encoderffmpegvorbis.h"
+#else
+#include "encoder/encodermp3.h"
+#include "encoder/encodervorbis.h"
+#endif
+#include "encoder/encoderwave.h"
+#include "encoder/encodersndfileflac.h"
+#include "encoder/encodermp3settings.h"
+#include "encoder/encodervorbissettings.h"
+#include "encoder/encoderwavesettings.h"
+#include "encoder/encoderflacsettings.h"
 
-Encoder::Encoder() {
+#include <QList>
+
+EncoderFactory EncoderFactory::factory;
+
+const EncoderFactory& EncoderFactory::getFactory()
+{
+    return factory;
+}
+
+EncoderFactory::EncoderFactory() {
+    // Add new supported formats here. Also modify the getNewEncoder/getEncoderSettings method.
+    m_formats.append(Encoder::Format("Wave PCM",ENCODING_WAVE, true));
+    m_formats.append(Encoder::Format("AIFF PCM",ENCODING_AIFF, true));
+    m_formats.append(Encoder::Format("Flac", ENCODING_FLAC, true));
+    m_formats.append(Encoder::Format("Mp3",ENCODING_MP3, false));
+    m_formats.append(Encoder::Format("Ogg Vorbis",ENCODING_OGG, false));
+}
+
+const QList<Encoder::Format> EncoderFactory::getFormats() const
+{
+    return m_formats;
 }
 
-Encoder::~Encoder() {
+Encoder::Format EncoderFactory::getSelectedFormat(UserSettingsPointer pConfig) const
+{
+    QString encoding = pConfig->getValueString(ConfigKey(RECORDING_PREF_KEY, "Encoding"));
+    foreach(Encoder::Format format, m_formats) {
+        if (format.internalName == encoding) {
+            return format;
+        }
+    }
+    qWarning() << "Format: " << encoding << " not recognized! Returning format WAV";
+    return m_formats.first();
+}
+Encoder::Format EncoderFactory::getFormatFor(QString formatText) const
+{
+    foreach(Encoder::Format format, m_formats) {
+        if (format.internalName == formatText) {
+            return format;
+        }
+    }
+    qWarning() << "Format: " << formatText << " not recognized! Returning format WAV";
+    return m_formats.first();
+}
+
+EncoderPointer EncoderFactory::getNewEncoder(Encoder::Format format,
+    UserSettingsPointer pConfig, EncoderCallback* pCallback) const
+{
+    EncoderPointer pEncoder;
+    if (format.internalName == ENCODING_WAVE) {
+        pEncoder = std::make_shared<EncoderWave>(pCallback);
+        pEncoder->setEncoderSettings(EncoderWaveSettings(pConfig, format));
+    } else if (format.internalName == ENCODING_AIFF) {
+        pEncoder = std::make_shared<EncoderWave>(pCallback);
+        pEncoder->setEncoderSettings(EncoderWaveSettings(pConfig, format));
+    } else if (format.internalName == ENCODING_FLAC) {
+        pEncoder = std::make_shared<EncoderSndfileFlac>(pCallback);
+        pEncoder->setEncoderSettings(EncoderFlacSettings(pConfig));
+    } else if (format.internalName == ENCODING_MP3) {
+        #ifdef __FFMPEGFILE__
```

Comment:
See the comment i put on top (with the #include): 
// TODO: Note: this is incomplete, since encoderffmpegmp3.cpp (and vorbis)
// are not compiled in any case.
I Just kept it as it was (moved it from enginerecord to here).
Also, i'm not sure if it really compiles currently, because i omitted the #ifdef and got a problem with the ffmpegresample.h header.
Finally, I believe that ffmpeg has an own implementation of most of the codecs it uses (even though it can be compiled to use existing codecs). As such, the quality is not the same than using "the real thing". 
Afaik, the ogg vorbis encoder used to be of lower quality, but maybe that has been improved. I'm not sure about its mp3 encoder.